solutions/usaco-969.mdx

```diff
@@ -277,3 +277,157 @@ public class pump{
 ```
 </JavaSection>
 </LanguageSection>
+
+## Alternative Implementation - Binary Search + DP
+
+Let $\texttt{dp}[i][j]$ represent the minimum path cost with a flow rate of $j$ going through junction $i$. In order to maintain a flow rate of $j$, the undirected edge going through $i$ must either have a flow of $j$ or there must be some previous edge with a flow of $j$. This leads us to the following transitions:
+
+ - $\texttt{dp}[i][j] = \min_{j \leq k \leq 1000}{(\texttt{dp}[i][j] + \texttt{cost}[i][x])}$, where $x$ represents an arbitrary junction connected to $i$ and $\texttt{cost}[i][x]$ represents the cost of the pipe connecting junction $i$ and junction $x$.
+ - $\texttt{dp}[i][k] = \min_{1 \leq k \leq j}{(\texttt{dp}[x][k] + \texttt{cost}[i][x])}$, where $x$ represents an arbitrary junction connected to $i$ and $\texttt{cost}[i][x]$ represents the cost of the pipe connecting junction $i$ and junction $x$.
+
+Similar to [Talent Show](http://www.usaco.org/index.php?page=viewproblem2&cpid=839), this problem asks us to find the maximum ratio multiplied by a large constant (in this case $10^6$) rounded down to the nearest integer, so we can binary search on the answer for this ratio!
+
+Let $R$ be the maximum ratio that we're trying to find multiplied by $10^6$. The given ratio $R$ works if $\frac{10^6 \cdot F}{C} \geq R$ is satisfied, where $F$ represents the flow rate in the path, and $C$ represents the total path cost in that path. This condition can more simply be written as $10^6 \cdot F - R \cdot C \geq 0$, where $C$ is $\texttt{dp}[n - 1][j]$ and $F$ is $j$, where the condition $1 \leq j \leq 10^6$ must hold.
+
+<LanguageSection>
+
+<CPPSection>
+	
+**Time Complexity:** $\mathcal{O}(N(N \cdot MAXF + M)\log MAXR)$
```

Comment:
this isn't supposed to pass :|

Reply:
My code doesn't pass or i have the wrong time complexity?

Reply:
`N*N*MAXF*log MAXR > 10^9`